File: Sources/ManageArmsCore/SkillManager.swift
import Foundation

/// スキルの有効化/無効化。DESIGN.md 3.2。
///
/// 有効/無効は全エージェント一括。Cursor と Codex は `~/.agents/skills/` を直読みするため、
/// エージェント別の on/off は原理的に不可能。
public enum SkillManager {

    public enum Failure: Error, Equatable, CustomStringConvertible {
        case notFound(String)
        case alreadyExists(String)
        case rollbackFailed(String)
        public var description: String {
            switch self {
            case .notFound(let n):
                String(localized: "\(n) が見つかりません")
            case .alreadyExists(let p):
                String(localized: "\(p) に既に別のものがあります。上書きしません")
            case .rollbackFailed(let message):
                String(localized: "操作に失敗し、元の状態も完全には復元できませんでした: \(message)")
            }
        }
    }

    /// 実体を置き場に戻し、Claude 用の symlink を張り直す。
    public static func enable(_ name: String, env: Environment, registry: inout Registry) throws {
        try ManagedLifecycle.enable(name, kind: .skill, env: env, registry: &registry)
    }

    /// Claude symlink を外し、実体を退避ディレクトリへ移す。**実体は消さない。**
    public static func disable(_ name: String, env: Environment, registry: inout Registry) throws {
        try ManagedLifecycle.disable(name, kind: .skill, env: env, registry: &registry)
    }

    /// Claude だけは共有ルートを読まないので symlink が要る（3.2）。
    /// Cursor / Codex は `~/.agents/skills` を直読みするため何もしない。
    static func linkForClaude(_ name: String, env: Environment, registry: Registry) throws {
        try ManagedLifecycle.link(name, kind: .skill, env: env, registry: registry)
    }
}

extension SkillManager {
    @discardableResult
    public static func removeExisting(_ url: URL, kind: Kind, project: String? = nil,
                                      env: Environment) throws -> URL? {
        try WriteGuard.assertUserArtifact(url, kind: kind, project: project, env: env)
        var result: NSURL?
        try FileManager.default.trashItem(at: url, resultingItemURL: &result)
        return result as URL?
    }

    /// 実体をゴミ箱へ移し、registry から外す。**完全削除しない** —
    /// 初心者が誤って消しても Finder から戻せる状態を残す（DESIGN.md 8 章）。
    /// 戻り値はゴミ箱に入った実体の位置（テストの後片付けに使う）。
    @discardableResult
    public static func remove(_ name: String, env: Environment, registry: inout Registry) throws -> URL? {
        try ManagedLifecycle.remove(name, kind: .skill, env: env, registry: &registry)
    }
}

/// Skill と Subagent に共通する guarded lifecycle。配置差だけを内側に持つ。
enum ManagedLifecycle {
    struct Layout {
        let store: URL
        let parked: URL
        let links: [URL]
    }

    static func layout(_ name: String, kind: Kind, env: Environment) -> Layout {
        if kind == .subagent {
            return Layout(
                store: env.agentStore.appending(path: "\(name).md"),
                parked: env.disabledAgentStore.appending(path: "\(name).md"),
                links: [env.home.appending(path: ".claude/agents/\(name).md"),
                        env.home.appending(path: ".cursor/agents/\(name).md")]
            )
        }
        return Layout(
            store: env.skillStore.appending(path: name),
            parked: env.disabledStore.appending(path: name),
            links: [env.claudeSkills.appending(path: name)]
        )
    }

    static func link(_ name: String, kind: Kind, env: Environment,
                     registry: Registry) throws {
        let fm = FileManager.default
        try WriteGuard.assertValidName(name)
        try Registry.assertReadable(env: env)
        let layout = layout(name, kind: kind, env: env)
        var previous: [URL: URL] = [:]
        for link in layout.links where WriteGuard.isSymlink(link) {
            try WriteGuard.assertMutable(link, env: env, registry: registry)
            if let target = WriteGuard.symlinkTarget(link) { previous[link] = target }
        }
        for link in layout.links where !WriteGuard.isSymlink(link)
            && fm.fileExists(atPath: link.path(percentEncoded: false)) {
            throw SkillManager.Failure.alreadyExists(link.path(percentEncoded: false))
        }
        var changed: [URL] = []
        do {
            for link in layout.links {
                try WriteGuard.assertSafeCreation(link, inside: link.deletingLastPathComponent(),
                                                  anchor: env.home)
                if WriteGuard.isSymlink(link) {
                    try fm.removeItem(at: link)
                }
                try fm.createDirectory(at: link.deletingLastPathComponent(),
                                       withIntermediateDirectories: true)
                try fm.createSymbolicLink(at: link, withDestinationURL: layout.store)
                changed.append(link)
            }
        } catch {
            let originalError = error
            do {
                for link in changed where WriteGuard.isSymlink(link) { try fm.removeItem(at: link) }
                for (link, target) in previous {
                    if !WriteGuard.isSymlink(link), !fm.fileExists(atPath: link.path) {
                        try fm.createSymbolicLink(at: link, withDestinationURL: target)
                    }
                }
            } catch let rollbackError {
                throw SkillManager.Failure.rollbackFailed("\(originalError); \(rollbackError)")
            }
            throw originalError
        }
    }

    static func enable(_ name: String, kind: Kind, env: Environment,
                       registry: inout Registry) throws {
        let fm = FileManager.default
        try WriteGuard.assertValidName(name)
        try Registry.assertReadable(env: env)
        let layout = layout(name, kind: kind, env: env)
        let originalRegistry = registry
        let wasParked = !fm.fileExists(atPath: layout.store.path)
        let previousLinks = Dictionary(uniqueKeysWithValues: layout.links.compactMap { link in
            WriteGuard.symlinkTarget(link).map { (link, $0) }
        })
        if !fm.fileExists(atPath: layout.store.path(percentEncoded: false)) {
            guard fm.fileExists(atPath: layout.parked.path(percentEncoded: false)) else {
                throw SkillManager.Failure.notFound(name)
            }
            try WriteGuard.assertMutable(layout.parked, env: env, registry: registry)
            try WriteGuard.assertSafeCreation(layout.store,
                                              inside: layout.store.deletingLastPathComponent(),
                                              anchor: env.home)
            try fm.createDirectory(at: layout.store.deletingLastPathComponent(),
                                   withIntermediateDirectories: true)
            try fm.moveItem(at: layout.parked, to: layout.store)
        }
        do {
            try link(name, kind: kind, env: env, registry: registry)
            var entry = registry.entry(named: name, kind: kind) ?? Registry.Entry(name: name, kind: kind)
            entry.disabled = false
            let fallback = entry
            registry = try Registry.update(env: env) { latest in
                var current = latest.entry(named: name, kind: kind) ?? fallback
                current.disabled = false
                latest.upsert(current)
            }
        } catch {
            let originalError = error
            registry = originalRegistry
            do {
                for link in layout.links where WriteGuard.isSymlink(link) {
                    if WriteGuard.symlinkTarget(link)?.standardizedFileURL == layout.store.standardizedFileURL {
                        try fm.removeItem(at: link)
                    }
                }
                for (link, target) in previousLinks
                    where !WriteGuard.isSymlink(link) && !fm.fileExists(atPath: link.path) {
                    try fm.createSymbolicLink(at: link, withDestinationURL: target)
                }
                if wasParked, fm.fileExists(atPath: layout.store.path),
                   !fm.fileExists(atPath: layout.parked.path) {
                    try fm.moveItem(at: layout.store, to: layout.parked)
                }
            } catch let rollbackError {
                throw SkillManager.Failure.rollbackFailed("\(originalError); \(rollbackError)")
            }
            throw originalError
        }
    }

    /// **前提の検査を全部先に済ませてから壊す。**
    /// 以前は symlink を先に外していたため、その後の `guard` で throw すると
    /// 「Claude からは消えたのに registry は有効のまま」という、
    /// 画面から追えない食い違いが残った（`Updater.apply` と同じ規律に揃える）。
    ///
    /// 破壊の順序は **実体の退避 → symlink の解除**。逆にすると、途中で失敗したときに
    /// 「リンクが無いだけ」の見えない状態になる。この順なら残るのはリンク切れで、
    /// 一覧が「読み込めません」として拾える（`ResourceRow.isUnusable`）。
    static func disable(_ name: String, kind: Kind, env: Environment,
                        registry: inout Registry) throws {
        let fm = FileManager.default
        try WriteGuard.assertValidName(name)
        try Registry.assertReadable(env: env)
        let layout = layout(name, kind: kind, env: env)
        let originalRegistry = registry

        guard fm.fileExists(atPath: layout.store.path(percentEncoded: false)) else {
            throw SkillManager.Failure.notFound(name)
        }
        try WriteGuard.assertMutable(layout.store, env: env, registry: registry)
        guard !fm.fileExists(atPath: layout.parked.path(percentEncoded: false)) else {
            throw SkillManager.Failure.alreadyExists(layout.parked.path(percentEncoded: false))
        }
        let links = layout.links.filter(WriteGuard.isSymlink)
        for link in links {
            try WriteGuard.assertMutable(link, env: env, registry: registry)
        }

        do {
            try WriteGuard.assertSafeCreation(layout.parked,
                                              inside: layout.parked.deletingLastPathComponent(),
                                              anchor: env.home)
            try fm.createDirectory(at: layout.parked.deletingLastPathComponent(),
                                   withIntermediateDirectories: true)
            try fm.moveItem(at: layout.store, to: layout.parked)
            for link in links { try fm.removeItem(at: link) }
            var entry = registry.entry(named: name, kind: kind) ?? Registry.Entry(name: name, kind: kind)
            entry.disabled = true
            let fallback = entry
            registry = try Registry.update(env: env) { latest in
                var current = latest.entry(named: name, kind: kind) ?? fallback
                current.disabled = true
                latest.upsert(current)
            }
        } catch {
            let originalError = error
            registry = originalRegistry
            do {
                if fm.fileExists(atPath: layout.parked.path),
                   !fm.fileExists(atPath: layout.store.path) {
                    try fm.moveItem(at: layout.parked, to: layout.store)
                }
                for link in links
                    where !WriteGuard.isSymlink(link) && !fm.fileExists(atPath: link.path) {
                    try fm.createDirectory(at: link.deletingLastPathComponent(),
                                           withIntermediateDirectories: true)
                    try fm.createSymbolicLink(at: link, withDestinationURL: layout.store)
                }
            } catch let rollbackError {
                throw SkillManager.Failure.rollbackFailed("\(originalError); \(rollbackError)")
            }
            throw originalError
        }
    }

    @discardableResult
    static func remove(_ name: String, kind: Kind, env: Environment,
                       registry: inout Registry) throws -> URL? {
        let fm = FileManager.default
        try WriteGuard.assertValidName(name)
        try Registry.assertReadable(env: env)
        let layout = layout(name, kind: kind, env: env)
        let originalRegistry = registry

        // `disable` と同じ理由で、検査を全部先に済ませてから壊す。
        let links = layout.links.filter(WriteGuard.isSymlink)
        let bodies = [layout.store, layout.parked]
            .filter { fm.fileExists(atPath: $0.path(percentEncoded: false)) }
        guard !bodies.isEmpty else { throw SkillManager.Failure.notFound(name) }
        for url in links + bodies {
            try WriteGuard.assertMutable(url, env: env, registry: registry)
        }

        var moved: [(original: URL, trash: URL)] = []
        do {
            for link in links { try fm.removeItem(at: link) }
            for url in bodies {
                var result: NSURL?
                try fm.trashItem(at: url, resultingItemURL: &result)
                if let trash = result as URL? { moved.append((url, trash)) }
            }
            registry = try Registry.update(env: env) { latest in
                latest.resources.removeAll { $0.name == name && $0.kind == kind.rawValue }
            }
        } catch {
            registry = originalRegistry
            do {
                for item in moved.reversed() where fm.fileExists(atPath: item.trash.path) {
                    try fm.moveItem(at: item.trash, to: item.original)
                }
                for link in links
                    where !WriteGuard.isSymlink(link) && !fm.fileExists(atPath: link.path) {
                    try fm.createDirectory(at: link.deletingLastPathComponent(),
                                           withIntermediateDirectories: true)
                    try fm.createSymbolicLink(at: link, withDestinationURL: layout.store)
                }
            } catch let rollbackError {
                throw SkillManager.Failure.rollbackFailed("\(error); \(rollbackError)")
            }
            throw error
        }
        return moved.last?.trash
    }
}

File: src/skillManager.ts
import { join } from "node:path";
import { KindId } from "./agent";
import { agentStore, claudeSkills, disabledAgentStore, disabledStore, Env, skillStore } from "./env";
import { AgentToolError } from "./errors";
import { Entry, Registry, upsert } from "./registry";
import * as guard from "./writeGuard";

/**
 * 有効化・無効化・削除。
 *
 * 有効/無効は全エージェント一括。Cursor と Codex は `~/.agents/skills/` を直読みするので、
 * エージェント別の on/off は原理的に不可能。
 */
export type Layout = {
  /** 実体の置き場。 */
  readonly store: string;
  /** 無効化したときの退避先。実体は消さない。 */
  readonly parked: string;
  /** 各エージェントから見えるようにするリンク。 */
  readonly links: string[];
  /** リンクの種別。Skill はディレクトリ、Subagent は単一ファイル。 */
  readonly linkKind: "dir" | "file";
};

export function layout(name: string, kind: KindId, env: Env): Layout {
  // MCP と Plugin はこちらの管理ストアに実体を持たない。ここを素通りさせると
  // `~/.agents/skills/<name>` を指してしまい、同名の Skill を消しにいく。
  if (kind !== "skill" && kind !== "subagent") {
    throw new AgentToolError("OPERATION_FAILED",
      `${kind} is managed by the agent, not by Agent Tool`);
  }
  if (kind === "subagent") {
    return {
      store: join(agentStore(env), `${name}.md`),
      parked: join(disabledAgentStore(env), `${name}.md`),
      // Subagent は共有ルートの慣習が無いのでリンクが 2 本要る。
      links: [join(env.home, ".claude", "agents", `${name}.md`),
              join(env.home, ".cursor", "agents", `${name}.md`)],
      linkKind: "file",
    };
  }
  return {
    store: join(skillStore(env), name),
    parked: join(disabledStore(env), name),
    // Claude だけは共有ルートを読まないのでリンクが要る。Cursor / Codex は直読み。
    links: [join(claudeSkills(env), name)],
    linkKind: "dir",
  };
}

const notFound = (name: string): never => {
  throw new AgentToolError("NOT_FOUND", `${name} was not found`);
};

const rollbackFailed = (original: unknown, rollback: unknown): never => {
  throw new AgentToolError("OPERATION_FAILED",
    `the operation failed and the original state could not be fully restored: ${original}; ${rollback}`);
};

const entryFor = (registry: Registry, name: string, kind: KindId): Entry =>
  registry.resources.find(item => item.name === name && item.kind === kind)
  ?? { name, kind, pinned: false, disabled: false };

/** 実体へのリンクを張り直す。既に別のものがある位置は上書きしない。 */
export function link(name: string, kind: KindId, env: Env, registry: Registry): void {
  guard.assertValidName(name);
  const plan = layout(name, kind, env);

  const previous = new Map<string, string>();
  for (const path of plan.links) {
    if (!guard.isLink(path)) continue;
    guard.assertMutable(path, env, registry);
    const target = guard.linkTarget(path);
    if (target !== null) previous.set(path, target);
  }
  for (const path of plan.links) {
    if (!guard.isLink(path) && guard.exists(path) && !guard.isManagedLink(path, plan.store)) {
      throw new AgentToolError("ALREADY_EXISTS", `${path} already holds something else; not overwriting`);
    }
  }

  const changed: string[] = [];
  try {
    for (const path of plan.links) {
      guard.prepare(path, join(path, ".."), env.home);
      if (guard.exists(path) || guard.isLink(path)) guard.removeLink(path);
      guard.createLink(plan.store, path, plan.linkKind);
      changed.push(path);
    }
  } catch (error) {
    try {
      for (const path of changed) {
        if (guard.isManagedLink(path, plan.store)) guard.removeLink(path);
      }
      for (const [path, target] of previous) {
        if (!guard.exists(path)) guard.createLink(target, path, plan.linkKind);
      }
    } catch (rollback) {
      rollbackFailed(error, rollback);
    }
    throw error;
  }
}

/** 実体を置き場に戻し、リンクを張り直す。 */
export function enable(name: string, kind: KindId, env: Env, registry: Registry): void {
  guard.assertValidName(name);
  const plan = layout(name, kind, env);
  const wasParked = !guard.exists(plan.store);

  if (wasParked) {
    if (!guard.exists(plan.parked)) notFound(name);
    guard.assertMutable(plan.parked, env, registry);
    guard.prepare(plan.store, join(plan.store, ".."),
      guard.isInside(plan.store, env.home) ? env.home : env.appSupport);
    guard.move(plan.parked, plan.store);
  }
  try {
    link(name, kind, env, registry);
    upsert(registry, { ...entryFor(registry, name, kind), disabled: false });
  } catch (error) {
    try {
      for (const path of plan.links) {
        if (guard.isManagedLink(path, plan.store)) guard.removeLink(path);
      }
      if (wasParked && guard.exists(plan.store) && !guard.exists(plan.parked)) {
        guard.move(plan.store, plan.parked);
      }
    } catch (rollback) {
      rollbackFailed(error, rollback);
    }
    throw error;
  }
}

/**
 * リンクを外し、実体を退避ディレクトリへ移す。実体は消さない。
 *
 * 前提の検査を全部先に済ませてから壊す。破壊の順序は実体の退避 → リンクの解除。
 * 逆にすると、途中で失敗したときに「リンクが無いだけ」の見えない状態が残る。
 * この順なら残るのはリンク切れで、一覧が「読み込めません」として拾える。
 */
export function disable(name: string, kind: KindId, env: Env, registry: Registry): void {
  guard.assertValidName(name);
  const plan = layout(name, kind, env);

  if (!guard.exists(plan.store)) notFound(name);
  guard.assertMutable(plan.store, env, registry);
  if (guard.exists(plan.parked)) {
    throw new AgentToolError("ALREADY_EXISTS", `${plan.parked} already holds something else; not overwriting`);
  }
  const links = plan.links.filter(path => guard.isManagedLink(path, plan.store));
  for (const path of links) guard.assertMutable(path, env, registry);

  try {
    guard.prepare(plan.parked, join(plan.parked, ".."),
      guard.isInside(plan.parked, env.home) ? env.home : env.appSupport);
    guard.move(plan.store, plan.parked);
    for (const path of links) guard.removeLink(path);
    upsert(registry, { ...entryFor(registry, name, kind), disabled: true });
  } catch (error) {
    try {
      if (guard.exists(plan.parked) && !guard.exists(plan.store)) {
        guard.move(plan.parked, plan.store);
      }
      for (const path of links) {
        if (!guard.exists(path)) guard.createLink(plan.store, path, plan.linkKind);
      }
    } catch (rollback) {
      rollbackFailed(error, rollback);
    }
    throw error;
  }
}

/**
 * リンクと実体を消し、registry から外す。
 * ゴミ箱へは送らないので、呼び出し前に必ず確認ダイアログを出す（設計決定 D-5）。
 */
export function remove(name: string, kind: KindId, env: Env, registry: Registry): void {
  guard.assertValidName(name);
  const plan = layout(name, kind, env);

  // `disable` と同じ理由で、検査を全部先に済ませてから壊す。
  const links = plan.links.filter(path => guard.isManagedLink(path, plan.store) || guard.isLink(path));
  const bodies = [plan.store, plan.parked].filter(guard.exists);
  if (bodies.length === 0) notFound(name);
  for (const path of [...links, ...bodies]) guard.assertMutable(path, env, registry);

  for (const path of links) guard.removeLink(path);
  for (const path of bodies) guard.remove(path);
  registry.resources = registry.resources.filter(item => !(item.name === name && item.kind === kind));
}

/**
 * 他のツールが入れた実体を消す。registry に載っていないので `assertMutable` は通らない。
 * 代わりに「既知ルート直下にあること」だけを条件にする（`assertUserArtifact`）。
 *
 * 同じ名前が複数のルートに現れる（実体 + 各エージェントへのリンク）ので、
 * 走査ホワイトリストの全ルートを見て一括で消す。消した位置を返す。
 */
export function removeUnmanaged(name: string, kind: KindId, env: Env): string[] {
  guard.assertValidName(name);
  if (kind !== "skill" && kind !== "subagent") {
    throw new AgentToolError("OPERATION_FAILED",
      `${kind} is managed by the agent, not by Agent Tool`);
  }
  const leaf = kind === "subagent" ? `${name}.md` : name;
  // リンク切れは `exists` が false になるので、リンクかどうかも見る。
  const targets = guard.userRoots(kind, env)
    .map(root => join(root, leaf))
    .filter(path => guard.exists(path) || guard.isLink(path));
  if (targets.length === 0) notFound(name);

  // `disable` と同じ理由で、検査を全部先に済ませてから壊す。
  for (const path of targets) guard.assertUserArtifact(path, kind, env);
  for (const path of targets) guard.remove(path);
  return targets;
}
